dora/core/client/fs/src/main/java/alluxio/client/file/DoraFileOutStream.java

```diff
@@ -0,0 +1,252 @@
+        if (mCanceled) {
+          mUnderStorageOutputStream.cancel();
+        } else {
+          mUnderStorageOutputStream.flush();
+          mUnderStorageOutputStream.close();
+          optionsBuilder.setUfsLength(mBytesWritten);
+        }
+      }
+
+      if (mAlluxioStorageType.isStore()) {
+        if (mCanceled) {
+          mNettyDataWriter.cancel();
+        } else {
+          // Note, this is a workaround to prevent commit(blockN-1) and write(blockN)
+          // race, in worse case, this may result in commit(blockN-1) completes earlier than
+          // write(blockN), and blockN evicts the committed blockN-1 and causing file lost.
+        }
+      }
+
+      // Whether to complete file with async persist request.
+      if (!mCanceled && mUnderStorageType.isAsyncPersist()
+          && mOptions.getPersistenceWaitTime() != Constants.NO_AUTO_PERSIST) {
+        optionsBuilder.setAsyncPersistOptions(
+            FileSystemOptionsUtils.scheduleAsyncPersistDefaults(
+                mContext.getPathConf(mUri)).toBuilder()
+                .setCommonOptions(mOptions.getCommonOptions())
+                .setPersistenceWaitTime(mOptions.getPersistenceWaitTime()));
+      }
+
+      // Complete the file if it's ready to be completed.
+      if (!mCanceled && (mUnderStorageType.isSyncPersist() || mAlluxioStorageType.isStore())) {
+        // record to metadata store
+        CompleteFilePOptions options = CompleteFilePOptions.newBuilder()
+            .setUfsLength(mNettyDataWriter.pos())//getBytesWritten())
+            .setCommonOptions(FileSystemMasterCommonPOptions.newBuilder().build())
+            .setContentHash("HASH-256") // compute hash here
+            .build();
+        mClosed = true;
+        mDoraClient.completeFile(mUri.toString(), options, mUuid);
+      }
+    } catch (Throwable e) { // must catch Throwable
+      throw mCloser.rethrow(e); // IOException will be thrown as-is.
+    } finally {
+      mClosed = true;
+      mCloser.close();
+    }
+  }
+
+  @Override
+  public void flush() throws IOException {
+    // TODO(yupeng): Handle flush for Alluxio storage stream as well.
+    if (mUnderStorageType.isSyncPersist()) {
+      mUnderStorageOutputStream.flush();
+    }
+  }
+
+  @Override
+  public void write(int b) throws IOException {
+    writeInternal(b);
+  }
+
+  @Override
+  public void write(byte[] b) throws IOException {
+    Preconditions.checkArgument(b != null, PreconditionMessage.ERR_WRITE_BUFFER_NULL);
+    writeInternal(b, 0, b.length);
+  }
+
+  @Override
+  public void write(byte[] b, int off, int len) throws IOException {
+    writeInternal(b, off, len);
+  }
+
+  private void writeInternal(int b) throws IOException {
+    if (mWriteToAlluxio) {
+      Integer intVal = b;
+      byte[] bytes = new byte[]{intVal.byteValue()};
+      mNettyDataWriter.writeChunk(bytes, 0, 1);
+    }
+
+    if (mUnderStorageType.isSyncPersist()) {
+      mUnderStorageOutputStream.write(b);
```

Comment:
I have different opinions on this way to do sync write.
In this PR, the client writes data to worker in Netty Writer, and at the same time to UFS from client.
Direct writing from client to UFS, instead of Worker writing to UFS, will consume more computing networks between clients and workers. But if we do writing from worker to UFS, it is using storage networks.
In typical data centers, the storage networks are more friendly on this and are dedicated for this.

Reply:
I figure this way the code is much easier to write.

@huanghua78 I imagine you are proposing something like an IO chain client -> worker -> UFS. I think your propose is making sense because:
1. The client doesn't have to write the same contents twice. This saves client(app) side CPU cycles. HOWEVER we do need to implement the client code carefully so we don't spin and wait for the worker to write to UFS.
2. By Alluxio being a cache, we expect more communication in client - worker than worker - UFS. I think we also assume the client-side resources (CPU/memory/network) is more the bottleneck? If that is the case then moving client side load to the worker makes sense.
3. On the worker side we might be able to optimize in the future (eg. by merging chunks and avoiding duplicates maybe?)

However, I don't think such chaining code exists and we are not very experienced about possible corner cases? So my guess at this moment is it's going to take much longer time (for example, take a look at how HDFS chains IO between 3 replicas). So it's unlikely to happen in this implementation?

Reply:
Let's leave this as a future work.